File: 1962-Remove-Stones-to-Minimize-the-Total/1962-Remove-Stones-to-Minimize-the-Total.py
class Solution:
    def minStoneSum(self, piles: List[int], k: int) -> int:
        heap = [-pile for pile in piles]
        heapq.heapify(heap)
        for i in range(k):
            largest_pile = heapq.heappop(heap)
            new_pile = largest_pile // 2
            heapq.heappush(heap, new_pile)
        total_stones = -sum(heap)
        return total_stones

#TC: O(n log n), where n is the number of piles, since we need to perform heap operations on each pile. 
#SC: O(n), since we need to store all the piles in the heap.

File: 1962-Remove-Stones-to-Minimize-the-Total/1962-Remove-Stones-to-Minimize-the-Total.cpp
class Solution {
public:
    int minStoneSum(vector<int>& piles, int k) {
        priority_queue<int> heap(piles.begin(),piles.end());
        for (int i = 0; i < k; i++) {
            int largest_pile = heap.top();
            heap.pop();
            int new_pile = largest_pile / 2;
            heap.push(largest_pile - new_pile);
        }
        int total_stones = 0;
        while (!heap.empty()) {
            total_stones += heap.top();
            heap.pop();
        }
        return total_stones;
    }
};
